MonoGame.Framework/Platform/Threading.cs

```diff
@@ -71,41 +102,84 @@ internal static void BlockOnUIThread(Action action)
             if (action == null)
                 throw new ArgumentNullException("action");
 
+            BlockOnUIThread(_metaAction, action);
+        }
+
+        /// <summary>
+        /// Runs the given action on the UI thread and blocks the current thread while the action is running.
+        /// If the current thread is the UI thread, the action will run immediately.
+        /// </summary>
+        /// <param name="action">The action to be run on the UI thread</param>
+        /// <param name="state">The data to pass to <paramref name="action"/></param>.
+        internal static void BlockOnUIThread<TState>(Action<TState> action, TState state)
+        {
+            if (action == null)
+                throw new ArgumentNullException("action");
+
 #if DIRECTX || PSM
-            action();
+            action(state);
 #else
             // If we are already on the UI thread, just call the action and be done with it
             if (IsOnUIThread())
             {
-                action();
+                action(state);
                 return;
             }
 
-            ManualResetEventSlim resetEvent = new ManualResetEventSlim(false);
-            Add(() =>
+            ManualResetEventSlim resetEvent = RentResetEvent();
+            var queuedAction = new StateActionHelper<TState>.QueuedAction
             {
-#if ANDROID
-                //if (!Game.Instance.Window.GraphicsContext.IsCurrent)
-                ((AndroidGameWindow)Game.Instance.Window).GameView.MakeCurrent();
-#endif
-                action();
-                resetEvent.Set();
-            });
-            resetEvent.Wait();
+                ResetEvent = resetEvent,
+                Action = action,
+                State = state
+            };
+
+            lock (_queuedActions)
+            {
+                StateActionHelper<TState>.Queue.Enqueue(queuedAction);
+                _queuedActions.Add(StateActionHelper<TState>.DequeueAction);
+            }
+
+            try
+            {
+                if (!resetEvent.Wait(MaxWaitForUIThread))
+                    throw new TimeoutException();
```

Comment:
I just updated MonoGame to latest develop and now I am having problems with this check, since the MaxWaitForUIThread time is too short in some cases.

Is there a specific reason for Threading class being internal and MaxWaitForUIThread a const, or would it be ok to make this class public and remove the const modifier so that the value could be changed?

Maybe just change the field static?

I can make a pull request if that change would be ok.

Reply:
Could you elaborate on the use case?

Reply:
Because of that check, some previously working code will crash, if for example in my case the update loop takes too long to finish during game startup and object creation.

I fixed my crashing by just increasing MaxWaitForUIThread to 10 seconds, and started thinking that maybe it would be good idea to allow users change the value however they wish to. Of course just by increasing the value would also solve the problem, but I see no reason why this value should be locked to something specific.

Reply:
Why would your startup take this much time? If you don't mind, I would like more context than just saying that it doesn't work for you because we might be missing something.

Usually, if doubling a timing value fixes an issue, it's not the right fix and it is likely that there's more to it.

Reply:
Well, several hours of debugging later I found out that I (maybe?) had older dependencies on my output folder, and those were causing the problems.
Redownloading  solved  my problems.

Sometimes rm -rf is actually a solution :D

Sorry about that...

Reply:
@roeyskoe I don't think that it was an issue from your end, I pushed an updated monogame to our testers and this crashes _a lot_. I believe that both the max wait, and the max thread count are an issue.

@TechnologicalPizza Do you have a suggestion to get rid of ```MaxWaitForUIThread``` and ```MaxPooledResetEvents``` entirely ? These can very frequently fail on simple texture loadings and I don't think that increasing them would fix anything.

I'm going to open up an issue and look into this today.

Reply:
Yeah, I just found out that the problem was actually garbage collector being forced to run on UI thread. 

I had one RenderTarget2D being turned into garbage and not disposed correctly. Then while creating a bunch of objects automatic GC starts, and since the Dispose function of GraphicsResource calls BlockOnUIThread, it may time out.

On my case I got it fixed it by just adding a call to _target.Dispose(), but it is not really ideal if you forget to do that.

Reply:
For the record #7471